### ISSUE TITLE
Create heading from anchor link binding

### ISSUE THREAD

**Issue by zdcthomas** [lervag/wiki.vim#191]
Hello! First off, I really love the feature set of the plugin, and I super appreciate all the work here! Thanks!

Second, I'm sorry if this isn't the right place for this, but I'm trying to put together a PR to add the ability to create a heading instead of another file when following a link to a heading, as well as another binding to create an anchor tag link. 

`I've been looking in autoload/wiki/link/_template.vim` for the link creation, but I'm getting a bit lost. Do you have any suggestions for where to look to get started? Thanks!

**Comment by lervag** [comment]
> Hello! First off, I really love the feature set of the plugin, and I super appreciate all the work here! Thanks!

Very glad to hear it :)

> Second, I'm sorry if this isn't the right place for this, but I'm trying to put together a PR ...

You've come to the right place! I'll be happy to help.

> ... to add the ability to create a heading instead of another file when following a link to a heading, as well as another binding to create an anchor tag link.

Ok, so, you want links like `[[#Some heading]]` to create the new heading in the same file if it does not already exist? Also, I'm not quite sure what you mean with "binding to create an anchor tag link".

> `I've been looking in autoload/wiki/link/_template.vim` for the link creation, but I'm getting a bit lost. Do you have any suggestions for where to look to get started? Thanks!

Definitely. So: When you follow a link, you typically use the `<cr>` mapping, i.e. `<plug>(wiki-link-follow)`. This runs `wiki#link#follow`, which first uses `wiki#link#get()` to parse the link under the cursor. This creates a "link object", which is more or less described by the `link/_template.vim` file, although I am not surprised to hear it is hard to understand. :p

Well, the idea is that a link object has the `.toggle()` method. If the link
has an url, then it should also have the `.follow()` method. The `.follow()`
method is tried first. And as you can see, `[[#...]]` is recognized as a `wiki`
link (`link/wiki.vim`) and it has a `.follow()` method defined by
`url/wiki.vim`. Thus, this is what is executed.

I think what you need to do is to add some kind of fallback; i.e., when `url/wiki.vim` tries to fallback on an anchor link in the same document, but if the anchor does not exist, then create it. See the `s:handler.follow()` and `s:handler.follow_anchor()` methods in `url/wiki.vim`. The code here is not so messy, but it may be somewhat hard to read.

So, to summarize, I believe you would need to modify the following code, somehow. But it would probably need some changes at other locations as well.

```vim
function! s:handler.follow_anchor() abort dict
  let l:old_pos = getpos('.')
  call cursor(1, 1)

  for l:part in split(self.anchor, '#', 0)
    let l:part = substitute(l:part, '[- ]', '[- ]', 'g')
    let l:header = '^#\{1,6}\s*' . l:part . '\s*$'
    let l:bold = wiki#rx#surrounded(l:part, '*')

    if !(search(l:header, 'Wc') || search(l:bold, 'Wc'))
      call setpos('.', l:old_pos)
      break
    endif
    let l:old_pos = getpos('.')
  endfor
endfunction
```

**Comment by zdcthomas** [comment]
This is fantastic! I super appreciate the detailed feedback and I'll try working through this soon. 

```
Also, I'm not quite sure what you mean with "binding to create an anchor tag link".
```
I just mean that in the same way that I can put my cursor over a word, hit `<cr>` and that word will be transformed into `[word](word.md)`, I want to also make a binding to transform a word into `[word](#word)`

I hope that makes sense!

Thanks again!

**Comment by lervag** [comment]
> This is fantastic! I super appreciate the detailed feedback and I'll try working through this soon.

Great, glad to hear you find it useful!

As you might understand, I care a lot about this plugin. But my time is quite limited, so helping others to improve it seems like a good use of my time. :)

> I just mean that in the same way that I can put my cursor over a word, hit `<cr>` and that word will be transformed into `[word](word.md)`, I want to also make a binding to transform a word into `[word](#word)`

Ah, I see. That would mean a new key binding, e.g. `<plug>(wiki-link-toggle-anchor)` with a corresponding command `:WikiLinkToggleAnchor` and a function `wiki#link#toggle_anchor`. The idea would be to use the "word" link, i.e. `link/word.vim`, and perhaps add a new type of toggle function for it.

> Thanks again!

No problem. I'll close the issue, but feel free to follow up the discussion. You can of course also open a work in progress PR with questions.

**Comment by zdcthomas** [comment]
Awesome! I'll keep working and let you know how it goes!
